### PR TITLE
Moved States tests to package "states" because actual state classes

### PR DESCRIPTION
**Pull request by yakisoba0**
have been moved to adequate package before.

Finished implementing player turns, but to completely close that case
Shot state needs to be implemented.

Message bus constructor has been made public due to state tests package
migration.

Introduced ResultState which will be activated when one of the players
will have no fleet available and to meet that condition Player and
Fleet classes needed to have introduced corresponding methods.

### CODE REVIEW

File: src/main/java/com/spanish_inquisition/battleship/server/game_states/ShotState.java [hunk 1]
```diff
@@ -13,9 +14,35 @@ public ShotState(Players players, MessageBus requestBus) {
...
+    private boolean didPlayerWon(Player player) {
+        return players.getOpponentOf(player).hasNoFleet();
+    }
+
+    private boolean shootIfPlayerSentValidMessage(Player player) {
+        if (requestBus.haveMessageFromSender(player.getPlayerId())) {
+            String shotMessage = getMessageContentFromPlayer(player);
+            if (shotMessage.contains(Header.MOVE_REGULAR.name())) {
+                //TODO: implement shoot event, notify players
+                players.switchCurrentPlayer();
```
Comment (mprtcz) [inline src/main/java/com/spanish_inquisition/battleship/server/game_states/ShotState.java]: 👍

File: src/main/java/com/spanish_inquisition/battleship/server/Player.java [hunk 2]
```diff
@@ -25,4 +25,8 @@ public void setFleet(Fleet fleet) {
     public Fleet getFleet() {
         return fleet;
     }
+
+    public boolean hasNoFleet() {
+        return fleet.hasNoShips();
+    }
```
Comment (arturAR) [inline src/main/java/com/spanish_inquisition/battleship/server/Player.java]: XD

File: src/test/java/com/spanish_inquisition/battleship/server/game_states/PlacingShipsStateTest.java [hunk 3]
```diff
@@ -74,17 +79,17 @@ public void testIfFleetsAreAssigned() {
...
-            Assert.assertEquals(1, playerShips.get(1).getShipPoints().size());
-            Assert.assertEquals(1, playerShips.get(2).getShipPoints().size());
-            Assert.assertEquals(1, playerShips.get(3).getShipPoints().size());
-            Assert.assertEquals(2, playerShips.get(4).getShipPoints().size());
-            Assert.assertEquals(2, playerShips.get(5).getShipPoints().size());
-            Assert.assertEquals(2, playerShips.get(6).getShipPoints().size());
-            Assert.assertEquals(3, playerShips.get(7).getShipPoints().size());
-            Assert.assertEquals(3, playerShips.get(8).getShipPoints().size());
-            Assert.assertEquals(4, playerShips.get(9).getShipPoints().size());
+            new SoftAssert().assertEquals(10, playerShips.size());
```
Comment (mprtcz) [inline src/test/java/com/spanish_inquisition/battleship/server/game_states/PlacingShipsStateTest.java]: I learned something from this 👍

File: src/main/java/com/spanish_inquisition/battleship/server/game_states/ShotState.java [hunk 4]
```diff
@@ -13,9 +14,35 @@ public ShotState(Players players, MessageBus requestBus) {
...
+        if(!shootIfPlayerSentValidMessage(currentPlayer)) {
+            return this;
         }
-        return null;
+        return !didPlayerWon(currentPlayer) ?
+                new PlayerActionState(players, requestBus) : new ResultState(players, requestBus);
+    }
+
+    private boolean didPlayerWon(Player player) {
+        return players.getOpponentOf(player).hasNoFleet();
```
Comment (arturAR) [inline src/main/java/com/spanish_inquisition/battleship/server/game_states/ShotState.java]: 👍

File: src/test/java/com/spanish_inquisition/battleship/server/game_states/ShotStateTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test
+    public void testShootMessageWithValidHeader() {
+        GameState ss = new ShotState(players, messageBus);
+        messageBus.addMessage(1,0, Header.MOVE_REGULAR.name()+":1;");
+        players.getOpponentOf(players.getCurrentPlayer()).setFleet(
+                new FleetBuilder().build(
+                        Header.FLEET_REQUEST.name() +
+                                ":[0,2,4,6,8,9,20,21,23,24,26,27,28,40,41,42,44,45,46,47];"
```
Comment (arturAR) [inline src/test/java/com/spanish_inquisition/battleship/server/game_states/ShotStateTest.java]: 🐽

**Pull request merged**